src/Request.cpp

```diff
@@ -449,10 +485,31 @@ auto Request::setCompletionStatus(
 #pragma GCC diagnostic pop
 }
 
-auto Request::onComplete() -> void
+auto Request::onComplete(EventLoop& event_loop, std::shared_ptr<SharedRequest> shared_request, bool response_wait_time_timeout) -> void
 {
-    RequestHandle request(&m_request_pool, std::unique_ptr<Request>(this));
-    m_on_complete_handler(std::move(request));
+    auto request_handle_ptr = RequestHandle(std::move(shared_request));
+    
+    // We only call the stored on complete function once!
+    if (!m_on_complete_has_been_called.exchange(true, std::memory_order_acquire))
+    {
+        // If the request did not time out and we have an iterator, remove it from the set so we won't try to call
+        // onComplete again.
+        if (!response_wait_time_timeout && m_response_wait_time_set_iterator.has_value())
+        {
+            event_loop.removeTimeoutByIterator(m_response_wait_time_set_iterator.value());
```

Comment:
are we sure we don't want to do this even if the on complete function has already been called? I think we went over this but checking just in case. Also if this is fine a comment about the lifetime here could be nice